vllm/hpu/ops.py

```diff
@@ -25,72 +25,247 @@ def gelu_new(output, input):
+                    with torch.profiler.record_function("repeat_interleave_seq_key_check_true"):
+                        # Handle MQA and GQA
+                        keys = torch.repeat_interleave(keys, num_queries_per_kv, dim=1)
+            with torch.profiler.record_function("gemm_attn_weight"):
+                q_bmm = query.reshape(1, num_seqs * num_query_heads, head_size).transpose(0, 1)
+                k_bmm = keys.reshape(num_seqs * num_query_heads, head_size, block_size)
+                attn_weights_2 = scale * torch.matmul(q_bmm, k_bmm).reshape(num_seqs, num_query_heads, 1, block_size)
+            # FIXME: need to restore attention masks in a more sensible way
+            if attn_masks is not None:
+                for j in range(num_seqs):  # can run in parallel
+                    with torch.profiler.record_function(f"seq_loop"):
+                        with torch.profiler.record_function("seq_context_len_fill"):
+                            seq_context_len[0] = context_lens.index_select(0, seq_index)[0]
+                        with torch.profiler.record_function("context_len_check"):
+                            if (block_index.mul(block_size)).lt(context_lens.index_select(0, seq_index)):
+                                with torch.profiler.record_function("context_len_check_true"):
+
+                                    attn_weight = torch.index_select(attn_weights_2, 0, seq_index).squeeze(0)
+                                    with torch.profiler.record_function("create_attn_mask_check"):
+                                        if attn_masks is not None:
+                                            with torch.profiler.record_function("create_attn_mask_check_true"):
+                                                attn_mask = torch.index_select(attn_masks.index_select(0, block_index), 0, seq_index)
+                                                attn_weight = torch.masked_fill(
+                                                    attn_weight, ~(attn_mask.unsqueeze(0).to(torch.bool)), torch.finfo(attn_weight.dtype).min
+                                                )
+
+                                    with torch.profiler.record_function("store_seq_weights"):
+                                        attn_weights.index_copy_(0, seq_index, attn_weight.unsqueeze(0))
+
+                        with torch.profiler.record_function("seq_index_inc"):
+                            seq_index.add_(1)
+            else:
+                attn_weights = attn_weights_2
+
+            with torch.profiler.record_function("store_block_weight"):
+                attn_weights_blocks.index_copy_(0, block_index, attn_weights.unsqueeze(0))
+
+        with torch.profiler.record_function("block_index_inc"):
+            block_index.add_(1)
+        if device == "hpu":
+            htorch.core.mark_step()
+
+    # Cleanup out-of-bound weights and values
+    attn_weights_blocks_filler = torch.finfo(query.dtype).min
```

Comment:
The "minimum value of query type" filler is also used elsewhere in this code (e.g. the `attn_weights_blocks` definition) and should probably be defined and commented on at the start of the function. The question also arises as to the type restrictions, since the required behavior for this filler is to produce a zero under `exp()`, and that behavior is not observed on FP16.